r1cs/gadgets/std/src/fields/cmp.rs

```diff
@@ -0,0 +1,261 @@
+use crate::{
+    boolean::Boolean,
+    fields::fp::FpGadget,
+    prelude::*,
+    ToBitsGadget,
+};
+use algebra::PrimeField;
+use r1cs_core::{ConstraintSystem, SynthesisError};
+use core::cmp::Ordering;
+
+impl<F: PrimeField> FpGadget<F> {
```

Comment:
Actually, for our applications, in which we compare only small numbers (max 64 bits in length), we can do better by defining the comparison directly for the UInt32/UInt64 gadgets and use them inside the circuits instead of reading FieldElement(s) out of the u64.
We can define a comparison trait and macroize the UInt gadgets like arkworks did to be more general and complete https://github.com/arkworks-rs/r1cs-std/blob/master/src/bits/uint.rs
And to prove an inequality of the from x >= y, where x and y are UInt gadgets, we can follow an approach similar to what we did for the ThresholdSignatureCircuit (enforcing number_of_valid_signatures >= threshold) https://github.com/HorizenOfficial/zendoo-sc-cryptolib/blob/development/doc/GenericThresholdCircuit.pdf, e.g. enforcing that pack(x_bits) - pack(y_bits) == pack(alloc(x_val-y_val).to_bits_with_length_restriction[..64])
(We can do this with 1 LC of the form 2^0(x_b0 - y_b0) + ... + 2^n-1(x_bn-1 - y_bn-1) = 2^0 diff_0 + ... + 2^n-1 * diff_n-1.
We can also integrate Add and Mul gadgets for UInt (we already have AddMany), and let them return an "overflow" bit that, if desired, we can enforce it to be 0. This will make operations and comparison completely controllable and secure.

Reply:
Fully agree with your proposal. See my final review comment.